### PR TITLE
fix(python)!: consistently use negative every as the default for offset in group_by_dynamic

### LINKED ISSUE #11163: group_by_dynamic may skip some entries if period != None
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
from datetime import datetime
import polars as pl
df = pl.DataFrame(
    {
        "time": pl.datetime_range(
            start=datetime(2021, 12, 16),
            end=datetime(2021, 12, 16, 3),
            interval="30m",
            eager=True,
        ),
        "n": range(7),
    }
)

print(df.group_by_dynamic('time', every='30m', closed='right').agg(pl.col('n')))
print(df.group_by_dynamic('time', every='30m', period='30m', closed='right').agg(pl.col('n')))

```

### Log output

_No response_

### Issue description

I get
```python
shape: (7, 2)
┌─────────────────────┬───────────┐
│ time                ┆ n         │
│ ---                 ┆ ---       │
│ datetime[μs]        ┆ list[i64] │
╞═════════════════════╪═══════════╡
│ 2021-12-15 23:30:00 ┆ [0]       │
│ 2021-12-16 00:00:00 ┆ [1]       │
│ 2021-12-16 00:30:00 ┆ [2]       │
│ 2021-12-16 01:00:00 ┆ [3]       │
│ 2021-12-16 01:30:00 ┆ [4]       │
│ 2021-12-16 02:00:00 ┆ [5]       │
│ 2021-12-16 02:30:00 ┆ [6]       │
└─────────────────────┴───────────┘
shape: (6, 2)
┌─────────────────────┬───────────┐
│ time                ┆ n         │
│ ---                 ┆ ---       │
│ datetime[μs]        ┆ list[i64] │
╞═════════════════════╪═══════════╡
│ 2021-12-16 00:00:00 ┆ [1]       │
│ 2021-12-16 00:30:00 ┆ [2]       │
│ 2021-12-16 01:00:00 ┆ [3]       │
│ 2021-12-16 01:30:00 ┆ [4]       │
│ 2021-12-16 02:00:00 ┆ [5]       │
│ 2021-12-16 02:30:00 ┆ [6]       │
└─────────────────────┴───────────┘
```

### Expected behavior

Both should give the top output, and `period` only control the length of the windows - not the starting point

I'd suggest changing

> offset of the window. If None and period is None it will be equal to negative every

to just 

> offset of the window. If None it will be equal to negative every

which would be more consistent / less surprising / easier to teach

### Installed versions

<details>

```

INSTALLED VERSIONS
------------------
commit           : 0f437949513225922d851e9581723d82120684a6
python           : 3.10.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.10.102.1-microsoft-standard-WSL2
Version          : #1 SMP Wed Mar 2 00:30:59 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 2.0.3
numpy            : 1.25.1
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 67.6.1
pip              : 23.1.2
Cython           : None
pytest           : 7.3.1
hypothesis       : 6.82.4
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.13.2
pandas_datareader: None
bs4              : 4.12.2
bottleneck       : None
brotli           : None
fastparquet      : 2023.7.0
fsspec           : 2023.6.0
gcsfs            : None
matplotlib       : 3.7.1
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 12.0.1
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.11.0
snappy           : None
sqlalchemy       : 2.0.20
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None
None

```

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #11163 

technically breaking, but to be honest I consider the current behaviour buggy as it's not reporting valid windows - my suggestion would be treat it as a bug fix

but, could deprecate the behaviour and make the change in 1.0 if people prefer

---

my understanding of `group_by_dynamic` is the windows are:
- [start, start + period)
- [start + every, start + every + period)
- [start + 2 every, start + 2 every + period)
- ...

and that `start` is `earliest_timestamp.truncate(every) + offset`

So, I don't really see why `offset` depends on `period` at all, but I might be missing something

### CODE REVIEW

File: py-polars/tests/unit/datatypes/test_temporal.py [hunk 1]
```diff
@@ -590,8 +590,13 @@ def test_truncate_negative_offset(tzinfo: ZoneInfo | None) -> None:
...
 
-        assert out.shape == (3, 4)
-        assert out["A"].to_list() == [["A", "A", "B"], ["B", "B", "B"], ["B", "C"]]
+        assert out.shape == (4, 4)
+        assert out["A"].to_list() == [
+            ["A"],
+            ["A", "A", "B"],
+            ["B", "B", "B"],
+            ["B", "C"],
+        ]
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/datatypes/test_temporal.py:599]: same problem as reported in the issue: the current behaviour is skipping out a window completely, i.e. the `[-2, 1)` one, which contains the first datapoint

File: py-polars/polars/lazyframe/frame.py [hunk 2]
```diff
@@ -2908,7 +2908,7 @@ def group_by_rolling(
         """
         index_column = parse_as_expression(index_column)
         if offset is None:
-            offset = f"-{_timedelta_to_pl_duration(period)}"
+            offset = _negate_duration(_timedelta_to_pl_duration(period))
```
Comment (ritchie46) [inline py-polars/polars/lazyframe/frame.py:2911]: Is this the root of the change? What did we do before and what do we do now?

Comment (MarcoGorelli) [reply py-polars/polars/lazyframe/frame.py:2911]: ah no, sorry, this is just a drive-by

the root of the change is in `group_by_dynamic`:
```diff
-            offset = f"-{_timedelta_to_pl_duration(every)}" if period is None else "0ns"
+            offset = _negate_duration(_timedelta_to_pl_duration(every))
```

Comment (ritchie46) [reply py-polars/polars/lazyframe/frame.py:2911]: Ah, right. Indeed confusing that we made that parameter dependent on the existence of period.

**Pull request merged**